c10/cuda/CUDAException.h

```diff
@@ -81,5 +94,152 @@ C10_CUDA_API void c10_cuda_check_implementation(
     const bool include_device_assertions);
 
+/// Holds information about any device-side assertions that fail.
+/// Held in managed memory and access by both the CPU and the GPU.
+struct DeviceAssertionData {
+  /// Stringification of the assertion
+  char assertion_msg[C10_CUDA_DEVICE_SIDE_MAX_STR_LEN];
+  /// File the assertion was in
+  char filename[C10_CUDA_DEVICE_SIDE_MAX_STR_LEN];
+  /// Name of the function the assertion was in
+  char function_name[C10_CUDA_DEVICE_SIDE_MAX_STR_LEN];
+  /// Line number the assertion was at
+  int line_number;
+  /// Number uniquely identifying the kernel launch that triggered the assertion
+  uint32_t caller;
+  /// block_id of the thread that failed the assertion
+  int32_t block_id[3];
+  /// third_id of the thread that failed the assertion
+  int32_t thread_id[3];
+};
+
+/// Used to hold assertions generated by the device
+/// Held in managed memory and access by both the CPU and the GPU.
+struct DeviceAssertionsData {
+  /// Number of assertions that have already been recorded in `assertions`
+  int assertion_count;
+  /// An array of assertions that will be written to in a race-free manner
+  DeviceAssertionData assertions[C10_CUDA_DEVICE_SIDE_ASSERTION_COUNT];
+};
+
+/// Use to hold info about kernel launches so that we can run kernels
+/// asynchronously and still associate launches with assertion failures
+struct CUDAKernelLaunchInfo {
+  /// Filename of the code where the kernel was launched from
+  std::string launch_filename;
+  /// Function from which the kernel was launched
+  std::string launch_function;
+  /// Line number of where the code was launched from
+  uint32_t launch_linenum;
+  /// Backtrace of where the kernel was launched from
+  std::string launch_stacktrace;
+  /// Kernel that was launched
+  std::string kernel_name;
+  /// Device the kernel was launched on
+  int device;
+  /// Stream the kernel was launched on
+  int32_t stream;
+  /// A number that uniquely identifies the kernel launch
+  uint32_t generation_number;
+};
+
+/// Circular buffer used to hold information about kernel launches
+/// this is later used to reconstruct how a device-side kernel assertion failure
+/// occurred CUDAKernelLaunchRegistry is used as a singleton
+class C10_CUDA_API CUDAKernelLaunchRegistry {
+ private:
+  /// Assume that this is the max number of kernel launches that might ever be
+  /// enqueued across all streams on a single device
+  static constexpr int max_kernel_launches = 1024;
+  /// Void deleter for UVM/managed memory pointers
+  static void uvm_deleter(DeviceAssertionsData* uvm_assertions_ptr);
+  /// How many kernel launch infos we've inserted. Used to ensure that circular
+  /// queue doesn't provide false information by always increasing, but also to
+  /// mark where we are inserting into the queue
+  uint32_t generation_number = 0;
+  /// Shared mutex between writer and accessor to ensure multi-threaded safety.
+  mutable std::mutex read_write_mutex;
```

Comment:
Good idea, thanks.

Reply:
Hmm, but if we do this then it might be that one thread's kernel has an assertion failure and then freezes. A second thread detects that the assertion failure has happened, but can't report on it because it does not have access to the first thread's `kernel_launches` vector.

Reply:
Hmm, I'm not sure I understand. Why does the kernel launched on the other thread need to know about an unrelated kernel that failed?

Reply:
I think the alternative is making the communication between the GPU and CPU more complex or expensive.

Reply:
I guess my point is, if we're crashing the cuda context, it doesn't really matter what happens in the other threads; we want to get the errors out and then crash the process.

Reply:
@ezyang - That's true, but the thread that initiates the crash (by discovering there was a problem) may not be the thread that has the kernel launch information associated with the failure.

This mutex is here because, iirc, we were worried that making `generation_number` atomic wouldn't be sufficient to prevent fast wrapping on the launch buffer.

I'd guess that multithreaded (as opposed to multiprocess) GPU handling is probably an edge case, so both the mutex and atomic will be low-contention.

Reply:
No, I don't think this is accurate. Things like torchdeploy and also C++ API are specifically for letting us do inferences while multithreading on one process.

Reply:
The benchmark below shows that the mutex doesn't affect speed.

Thinking about it more, if each thread allocated it's own assertion failure buffer on each GPU it interacted with, that would also solve the problem and eliminate the mutex. But that would also use (on the machine I'm working on) 80x more memory.

The current on-device memory usage is ~15kB. 80x that would be 1.2MB, which is still probably just fine on today's multigigabyte GPUs.

Still, it might be better to get this landed so folks can begin using it and switch to separate buffers in a follow-up.